core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -239,10 +233,89 @@ private static MapperParsingException wrapInMapperParsingException(SourceToParse
         return new MapperParsingException("failed to parse", e);
     }
 
-    static ObjectMapper parseObject(ParseContext context, ObjectMapper mapper, boolean atRoot) throws IOException {
+    /** Creates a Mapping containing any dynamically added fields, or returns null if there were no dynamic mappings. */
+    static Mapping createDynamicUpdate(Mapping mapping, DocumentMapper docMapper, List<Mapper> dynamicMappers) {
+        if (dynamicMappers.isEmpty()) {
+            return null;
+        }
+        // We build a mapping by first sorting the mappers, so that all mappers containing a common prefix
+        // will be processed in a contiguous block. When the prefix is no longer seen, we pop the extra elements
+        // off the stack, merging them upwards into the existing mappers.
+        Collections.sort(dynamicMappers, (Mapper o1, Mapper o2) -> o1.name().compareTo(o2.name()));
+        List<ObjectMapper> parentMappers = new ArrayList<>();
+        // create an empty root object which updates will be propagated into
+        RootObjectMapper.Builder rootBuilder = new RootObjectMapper.Builder(docMapper.type());
+        RootObjectMapper.BuilderContext context = new RootObjectMapper.BuilderContext(Settings.EMPTY, new ContentPath());
+        parentMappers.add(rootBuilder.build(context));
```

Comment:
I wish the root of the mapping update was created with a call to clone() so that we can be sure to not ovverride the configuration of the root object with defaults?